utils/attribution.py

```diff
@@ -0,0 +1,226 @@
+from jaxtyping import Float
+from typing import Optional, Union, List, Literal, Tuple
+import torch as t
+from torch import Tensor
+from transformer_lens import HookedTransformer, ActivationCache
+import einops
+from circuitsvis.attention import attention_heads, attention_patterns
+
+
+def concat_lists(*lists):
+    return [x for l in lists for x in l]
+
+
+def diagonalise(tensor: Tensor, dim: int):
+    '''
+    Returns a tensor with dimension `dim` duplicated, diagonal elements along these two dims
+    are given by `tensor`, and off-diags are zero.
+    '''
+    shape = tensor.shape
+    new_shape = shape[:dim] + (shape[dim],) + shape[dim:]
+    new_tensor = t.zeros(new_shape, dtype=tensor.dtype, device=tensor.device)
+    indices = concat_lists(
+        [slice(None)] * dim,
+        [range(shape[dim]), range(shape[dim])],
+        [slice(None)] * (len(shape) - dim - 1)
+    )
+    new_tensor[indices] = tensor
+    return new_tensor
+
+
+def from_cache(
```

Comment:
not a very descriptive name!

Reply:
also this is printing too much out too much unlabelled stuff. Add verbose argument?

Reply:
It's hard to read too, can you split up into separate functions?